File: backend/src/main/kotlin/com/example/backend/graphql/GraphQLInstantCoercing.java
package com.example.backend.graphql;

import java.math.BigInteger;
import java.time.Instant;

import graphql.language.IntValue;
import graphql.schema.Coercing;

public class GraphQLInstantCoercing implements Coercing<Instant, Long> {

	@Override
	public Long serialize(Object input) {
		if (input instanceof Instant) {
			return ((Instant) input).toEpochMilli();
		}
		return null;
	}

	@Override
	public Instant parseValue(Object input) {
		if (input instanceof Long) {
			return Instant.ofEpochMilli((Long) input);
		}
		return null;
	}

	@Override
	public Instant parseLiteral(Object input) {
		if (!(input instanceof IntValue)) {
			return null;
		}

		BigInteger instant = ((IntValue) input).getValue();
		return this.parseValue(instant.longValueExact());
	}
}

File: backend/src/main/kotlin/com/example/backend/graphql/GraphQLInstantCoercing.kt
package com.example.backend.graphql

import java.time.Instant
import graphql.language.IntValue
import graphql.schema.Coercing

class GraphQLInstantCoercing : Coercing<Instant?, Long?> {
    override fun serialize(input: Any): Long? {
        if (input is Instant) {
            return input.toEpochMilli()
        }
        return null
    }

    override fun parseValue(input: Any): Instant? {
        if (input is Long) {
            return Instant.ofEpochMilli(input)
        }
        return null
    }

    override fun parseLiteral(input: Any): Instant? {
        if (input !is IntValue) {
            return null
        }

        val instant = input.value
        return this.parseValue(instant.longValueExact())
    }
}
